### PR TITLE
Sprite Builder

### PR DESCRIPTION
**Pull request by JacobBovee**
Features added in this PR:
- Responsive Sprite Editor with scaling
- Simulator display with position picker
- Resizable modal based on screen size
- Block highlighting
- Text inputs for sliders and position pickers
- General responsive fixes included scrollable modal and modal overlay.  

TODO's have been included in this pull request as I would like to fix a number of bugs after my code has been reviewed. Going forward I would also like to run our accessibility tool over our code and make all changes recommended, as well as building out a richer question set which will be included in the "pxt-arcade" PR.

sprite-wizard

### CODE REVIEW

**Comment by pelikhan** [comment]
Please add a custom build of the editor so that we can try out the feature.

**Comment by riknoll** [comment]
@JacobBovee ping me with the branches and I'll create a build for you

**Comment by pelikhan** [comment]
Whatever you use to test your code.

File: webapp/src/positionPicker.tsx [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+        const { x, y } = this.getScaledPoints();
+
+        return {
+            x: this.unScalePoint(x, true),
+            y: this.unScalePoint(y),
+        };
+    }
+
+    protected getDotPosition() {
+
```
[the remark this answers is not in the document]
Comment (JacobBovee) [reply webapp/src/positionPicker.tsx]: This I fixed thank you.

File: webapp/src/positionPicker.tsx [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+    }
+
+    protected getScaledPoints() {
+        const { x, y } = this.state;
+        console.log('Scale =>', this.state.scale);
```
[the remark this answers is not in the document]
Comment (JacobBovee) [reply webapp/src/positionPicker.tsx]: Yeah you're right I removed it thank you.

File: webapp/src/positionPicker.tsx [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+        this.setScale();
+        window.addEventListener('resize', this.setScale);
+    }
+
+    /**
+     * Sets the number to scale the position picker and simulator display
+     */
+    protected setScale() {
+        // 1023 - full size value
+        const height = window.innerHeight;
```
[the remark this answers is not in the document]
Comment (JacobBovee) [reply webapp/src/positionPicker.tsx]: Added constants

File: webapp/src/simulatorDisplay.tsx [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+export class SimulatorDisplay extends data.Component<SimulatorDisplayProps, {}> {
+
+    protected scalePixel(num: number, noPx?: true) {
+        const { scale } = this.props;
+        return `${num * scale}${!noPx && 'px'}`;
+    }
+
+    renderCore() {
+        const { children, scale } = this.props;
+
```
Comment (darzu) [inline webapp/src/simulatorDisplay.tsx]: Add a comment mentioning how you compute this raw SVG

File: webapp/src/positionPicker.tsx [hunk 5]
```diff
@@ -0,0 +1,241 @@
+import * as React from 'react';
```
Comment (darzu) [inline webapp/src/positionPicker.tsx]: does it have to be in webapp/src/ ? spriteEditor has a folder in pxtlib I think. Preferably this should go in a folder with other snippet builder specific stuff

File: webapp/src/positionPicker.tsx [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+        // Run once on component mount
+        this.setScale();
+        window.addEventListener('resize', this.setScale);
+    }
+
+    /**
+     * Sets the number to scale the position picker and simulator display
+     */
+    protected setScale() {
+        // 1023 - constant (FULLSIZE_BROWSER_HEIGHT)
```
Comment (darzu) [inline webapp/src/positionPicker.tsx]: what are these comments about?

File: webapp/src/simulatorDisplay.tsx [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+                            height="120"
+                            className="borders"
+                            style={{
+                                transform: `scale(${scale})`,
+                                transformOrigin: `0 0`,
+                                height: 229,
+                                width: 312,
+                                left: this.scalePixel(28),
+                                top: this.scalePixel(20),
+                                clipPath: `polygon(0px 5.33333px, 5.33333px 0px, 338.667px/1.2) 0px, 344px 5.33333px, 344px 256.667px, 338.667px 262px, 5.33333px 262px, 0px 256.667px)`,
```
Comment (darzu) [inline webapp/src/simulatorDisplay.tsx]: Any time you have a bunch of fixed values, try to provide a comment and some insight into how they were picked

File: webapp/src/snippetBuilder.tsx [hunk 8]
```diff
@@ -42,6 +42,8 @@ interface SnippetBuilderState {
  * An output type is attached to the start of your markdown allowing you to define a number of markdown output. (blocks, lang)
  * An initial output is set and outputs defined at each questions are appended to the initial output.
  * answerTokens can be defined and are replaced before being outputted. This allows you to output answers and default values.
+ * TODO:
+ * 1. Richer questions for different sprite paths
```
Comment (darzu) [inline webapp/src/snippetBuilder.tsx]: Maybe elaborate this a bit so non-you devs can understand this TODO :)

File: webapp/src/snippetBuilder.tsx [hunk 9]
```diff
@@ -153,18 +166,56 @@ export class SnippetBuilder extends data.Component<SnippetBuilderProps, SnippetB
...
         const { config, tsOutput } = this.state;
+
         // Attaches starting and ending line based on output type
         let md = `\`\`\`${config.outputType}\n`;
-        md += this.replaceTokens(tsOutput);
+        md += this.replaceTokens(this.highlightEditedBlocks(tsOutput));
         md += `\n\`\`\``;
-
+        // Removeswhitespace
+        // TODO(jb) md.replace(/\s/g, '_');
```
Comment (darzu) [inline webapp/src/snippetBuilder.tsx]: maybe state what's wrong so if you don't get to this, we can

File: webapp/src/spriteEditor.tsx [hunk 10]
```diff
@@ -12,120 +12,162 @@ interface ISpriteEditorProps {
...
-                    .replace(imgTag, '')
-                    .replace(endQuote, '')
+            .then((blocksInfo) => {
+                this.blocksInfo = blocksInfo;
+                this.renderSpriteEditor();
+            });
+    }
+
+    protected setScale() {
+        // 1023 - full size value
```
Comment (darzu) [inline webapp/src/spriteEditor.tsx]: more magic unexplained constants

**Comment by darzu** [comment]
> @JacobBovee ping me with the branches of arcade, common packages, etc, and I'll create a build for you

@riknoll , the private build doesn't seem to have his CSS changes. Is this expected?

File: webapp/src/positionPicker.tsx (hunk 6 above)
Comment (JacobBovee) [reply webapp/src/positionPicker.tsx]: It's the scale = 1 height and width

File: webapp/src/spriteEditor.tsx (hunk 10 above)
Comment (JacobBovee) [reply webapp/src/spriteEditor.tsx]: I replaced these with real constants

**Pull request merged**